src/interceptors/ClientRequest/NodeClientRequest.ts

```diff
@@ -99,6 +99,11 @@ export class NodeClientRequest extends ClientRequest {
   }
 
   write(...args: ClientRequestWriteArgs): boolean {
+    // Fallback to native behavior to throw 
+    if (this.destroyed || this.finished) {
```

Comment:
I believe neither of these properties will be set immediately after calling `.end()` for bypassed requests because they are set by Node in `super.end()` which is called asynchronously. 

Given the current limitations, I think the most straightforward way to fix this is to have an internal `this.requestSent` boolean flag that will control whether `.write()` calls should be buffered anymore. We can set that flag to `true` immediately in `.end()` regardless of the request listener lookup result. 

```js
write(...) {
  if (this.requestSent) {
    throwNodeWriteAfterEndError()
  }
}

end(...) {
  // Mark the request as sent immediately so
  // the interceptor would know to stop buffering
  // ".write()" calls after this point: that's a no-op.
  this.requestSent = true
}
```

What do you think about this?

I dislike introducing any new state but I don't suppose we can rely on request properties since those we need are set async anyway:

- `finished` is set by `respondWith()` in case of mock responses.
- `finished` is set by Node by `super.end()` in case of bypassed responses. 

And both happen asynchronously after the request listener promise settles.

Reply:
I agree! I dislike this as well.

We can alternate the `finished` property:

```
end(..) {
  this.finished = true

  until(...).then(() => {
    this.finished = false
  })
}
```

But I think this is more confusing and error-prone than a new state.

Reply:
And we'd likely have to recreate the original Node.js error here because calling `super.write()` in this case won't produce the "can't write after end error". It will very likely produce something connection-related because the connection has failed to establish. 

It shouldn't be hard replicating the Node's native error.